Enterprise Fees With Tax Report By Producer

abdellani opened the pull request:
#### What? Why?

- Closes #6844

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->



#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

- described on https://github.com/openfoodfoundation/openfoodnetwork/issues/6844
- 

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category: Technical changes

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
<!-- Does this PR depend on another one?
     Add the link or remove this section. -->



#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
     List them here or remove this section. -->

abdellani commented on app/services/order_fees_handler.rb:
```diff
@@ -1,12 +1,12 @@
 
   def initialize(order)
     @order = order
-    @distributor = order.distributor
```
I think it is not necessary to store a reference to the `order_cycle` and the `distributor` if we can delegate those two to the order.

Also, it was a source of problems when I was writing the tests.
 
```ruby
create(:simple_order_cycle)
create(:order_with_distributor, distributor: distributor)  # order created fees_handler.order_cycle = nil
order.update!({order_cycle_id: order_cycle.id,ship_address_id: ship_address.id})
order.reload
order.recreate_all_fees! # this will not work, because the OrderFeesHandler was initialized with  order_cycle= nil

```

abdellani commented:
The PR is not done yet. I didn't implement the filters yet.

I want to have feedback about the current implementation because It's clear to me that there are a lot of database queries that will make the report generation slow.
Is this acceptable?

@filipefurtad0 
Can you please check if the implemented tests look good?

mkllnk commented on lib/reporting/reports/enterprise_fee_summary/enterprise_fees_with_tax_report_by_producer.rb:
```diff
@@ -0,0 +1,292 @@
+module Reporting
+  module Reports
+    module EnterpriseFeeSummary
+      class EnterpriseFeesWithTaxReportByProducer < ReportTemplate
```
I think we need to add to the Rubocop `_todo` list. There's not easy way of complying to the styles here. It's part of the current report architecture to have those long methods. It would be good to fix that at some point.

mkllnk reviewed (changes requested):
Looking good but one file needs renaming.

abdellani commented:
@lin-d-hop, @filipefurtad0 

I have some questions:

- Do we list the coordinator fees in this report? 
- How about the enterprise fees calculated per order?
- I think the parent report `Enterprise Fee Summary` should be renamed `Enterprise Fee` since it includes more than the summary now. Is that ok?
![image](https://user-images.githubusercontent.com/16113997/215456287-f2d6f9ab-b7cc-4108-bff4-22aefc47e2ef.png)

abdellani commented:
I noticed an issue in the class `Reporting::LineItems`.
`Reporting::LineItems#orders` returns all the orders without applying the `supplier_id` filter.

https://github.com/openfoodfoundation/openfoodnetwork/blob/d5ae5c9af41cd814a0da0abf4bc83d61dab542e1/lib/reporting/line_items.rb#L13-L25

rioug commented on lib/reporting/reports/enterprise_fee_summary/enterprise_fees_with_tax_report_by_producer.rb:
```diff
@@ -0,0 +1,354 @@
+              adjustable_type: "Spree::Adjustment"
+            ).pluck(:originator_id)
+
+            tax_rate_ids << nil if tax_rate_ids.empty?
```
I don't really understand why we are doing this ? is that because `join_tax_rate` needs to return a non empty hash ?

abdellani replied:
We are grouping orders by `[entreprise_rate, tax_rate, supplier_id, distributor_id and order_cycle_id]`

if the order doesn't match any `tax_rate`, we want to have  `nil` in the grouping key:  `[entreprise_rate, nil, supplier_id, distributor_id and order_cycle_id]`

abdellani replied:
Yes, it's like a SQL left join between `LineItems` X `Tax rates`

rioug reviewed (commented):
Looks good to me code wise. I did leave a comment, as something wasn't quite clear to me .
I lack context to understand it all so I'll defer to others to make sure it does what it is supposed to do.

mkllnk commented on lib/reporting/reports/enterprise_fee_summary/enterprise_fees_with_tax_report_by_producer.rb:
```diff
@@ -0,0 +1,354 @@
+        end
+
+        def producer(query_result_row)
+          Enterprise.where(id: supplier_id(query_result_row)).pluck(:name).first
```
I just learned about `pick` which applies `limit 1` as well:
```suggestion
          Enterprise.where(id: supplier_id(query_result_row)).pick(:name)
```

mkllnk commented on app/views/admin/reports/filters/_enterprise_fee_summary.html.haml:
```diff
@@ -1,31 +1 @@
-  .alpha.two.columns= label_tag nil, t(:report_payment)
-  .omega.fourteen.columns
-    = collection_select(:q, :payment_method_ids, @report.permissions.allowed_payment_methods, :id, :name, {selected: params.dig(:q, :payment_method_ids)}, {class: "select2 fullwidth", multiple: true})
+= render 'admin/reports/date_range_form', f: f
```
You got some additional whitespace here.

abdellani replied:
are you talking about the space after the comma `,` or `render`?

mkllnk replied:
The last line of the file. Github doesn't display that properly here.

mkllnk commented on lib/reporting/reports/enterprise_fee_summary/enterprise_fees_with_tax_report_by_producer.rb:
```diff
@@ -1,17 +1,15 @@
 # frozen_string_literal: true
 
+# rubocop:disable Metrics/ClassLength
```
I would have added this to the todo file because I would like to resolve this problem one day.

mkllnk replied:
Sorry, I meant the `.rubocop_todo.ym` file.

mkllnk reviewed (approved):
The code looks good. It's complicated but I don't know if there's a better way of writing this. I also didn't assess the performance but I guess that we can improve that later. Can you fill in the PR description?

dacook commented on lib/reporting/reports/enterprise_fee_summary/enterprise_fees_with_tax_report_by_producer.rb:
```diff
@@ -88,7 +125,10 @@ def join_tax_rate
-            order.line_items.map(&:supplier_id).map do |supplier_id|
+            order.line_items
+              .map(&:supplier_id)
+              .filter(&method(:supplier_ids_filter))
```
I can't help but wonder if we could filter in the main `LineItems` query, which I imagine to be much  more efficient. But I don't know if that would work or not anyway ;)

dacook replied:
Today I learned about `&method()`, that's really handy thanks!

abdellani replied:
@dacook 
In `query_result`, I'm trying to do what looks like a left join between the tables:

`Orders` x `Enterprise Fees` x `Tax Rates` x `Suppliers`

the result will be an array of hashes 
```ruby
[entreprise_rate, tax_rate, supplier_id, distributor_id and order_cycle_id] => order
```

Now, if you check the initial set of `orders`, you'll find that's calculated from the `line items`
```ruby
orders = report_line_items.list.map(&:order).uniq
```
Those line items are already filtered by `supplier_id` (as you suggested on your comment).

abdellani replied:
Now, If I don't filter again on this method. The line items that were filtered on the main query will appear again : (
```ruby
            order.line_items
              .map(&:supplier_id)
```
this is why I added `.filter(&method(:supplier_ids_filter))`.

dacook reviewed (changes requested):
Great, a nice big powerful report.

Looks mostly good to me. I had a few comments about arranging commits, I hope they make sense. No need to change them now though.

I've added a commit to demonstrate the rubocop file. I also wonder about the efficiency of the report, but we can see how it goes.

`±` The only change is that I noticed a spelling mistake in a method name.

abdellani commented on lib/reporting/reports/enterprise_fee_summary/enterprise_fees_with_tax_report_by_producer.rb:
```diff
@@ -1,17 +1,15 @@
-          @parameters = Parameters.new(p || {})
-          @parameters.validate!
           @permissions = Permissions.new(user)
-          @parameters.authorize!(@permissions)
```
When I used `git rebase -i HEAD~X`, I was trying to keep the history of the PR clean.
This also makes it difficult for me to keep track of some changes.
Sorry about that. From now, I'll avoid `rebase` after getting a review (especially on big PRs)

dacook replied:
I'm not saying to never do a `rebase`, in fact my point is to keep the git history clean before requesting a review. (but carefully, because you can fall into a hole, which has happened to me a lot).

As you say, I think it's usually good to avoid the rebase after a review, so that we can discuss and review the iterations.

abdellani commented on lib/reporting/reports/enterprise_fee_summary/enterprise_fees_with_tax_report_by_producer.rb:
```diff
@@ -0,0 +1,292 @@
+                  .where(adjustable_id: enterprise_fee_adjustment_ids)
+                  .pluck("sum(amount)")
+                  .first || 0
+                total_excl_tax = enterprise_fees.sum(&:second) - included_tax
```
I stuck for a while to answer this feedback : )

Some enterprise fees are connected to `line items` and others are connected to `orders`.
To load all the enterprise fee tax related to an order, I need to use:
```ruby 
order.all_adjustments.tax.inclusive
            .where(adjustable_type: 'Spree::Adjustment')
            .where(adjustable_id: enterprise_fee_adjustment_ids_for_orders(order_ids))
```
If you check the methods used in this report to calculate the values for the columns, you'll see that I didn't load `order.all_adjustments`. 

Instead, I used `Spree::Adjustment.tax` to do the calculation on the database and fetch directly the result without loading all the data. For example:
```ruby
        def tax(query_result_row, all: false, included: nil)
          order_ids = orders(query_result_row).map(&:id)
          adjustment_ids = enterprise_fee_adjustemnt_ids(query_result_row)
          query = Spree::Adjustment.tax
          query = query.where(included: true) unless included.nil?
          query = query.where(originator_id: tax_rate_id(query_result_row)) unless all == true
          query = query.where(order_id: order_ids)
            .where(adjustable_type: 'Spree::Adjustment')
            .where(adjustable_id: adjustment_ids)
            .pluck("sum(amount)")

          query.first || 0
        end
```
This will run two  queries (if we don't count the query used to load the order in the main query):
1. one query to get adjustment_ids 
2. one query to get tax (after being already calculated)

In the other approach, I would need to run order.all_adjustments.tax.where(...) for every order which would require:
1. running more queries (1 query per order)
2. loading more data 
    a. will need more RAM
    b. query will be slower to fetch data
    c. calculation will be slower (will run on ruby) 
 

I tried the other approach, the code will look like 
```ruby
        def included_tax_for_order_ids(orders)
          enterprise_fee_adjustment_ids = enterprise_fee_adjustment_ids_for_orders(orders.map(&:id))
          orders.map do |order|
            order.all_adjustments.tax.inclusive.filter do |adjustment|
              adjustment.adjustable_type== 'Spree::Adjustment' &&
              enterprise_fee_adjustment_ids.includes?(adjustment.adjustable_id) 
            end.sum(&:amount)
          end.sum
```

What do you think?

dacook replied:
Wow, this is quite complex. I thought that the summary row would only show a simple aggregation of the data that is  already loaded in the line items, but it sounds like that is not the case.

I definitely don't want to suggest a solution that uses more memory and more queries, so it looks like your first solution is the best. Thanks for the thorough explanation!

abdellani replied:
> I thought that the summary row would only show a simple aggregation of the data that is already loaded in the line items

This is correct but to avoid loading everything into the memory, I pushed the work to the database.

dacook replied:
Oh, I believe that all `LineItems` are still in memory when the summary is created (which can be quite a lot, but is a necessary result of this framework that we use). 
So I don't think directly using the `LineItem` objects to produce the aggregate is going to use any more memory, and would probably be as fast as using the DB to aggregate. But there might be more to this equation, and I don't think it's necessary to deliberate any further. This still works :D

abdellani replied:
> I believe that all LineItems are still in memory when the summary is created

Yes, that's correct. the `LineItems` are loaded, but:

1. The data we're looking for is stored on the `Adjustment` model.  (but I didn't need to load them) 
2. All the `Adjustments` have a reference to the `order`, but not all the adjustment refers to the `LineItems`.

The second point means that we need to load the orders and from there we can load the adjustments.

dacook replied:
Ah, thanks for explaining further :)

dacook reviewed (approved):
👍

dacook commented:
I wasn't sure if the testing notes above were completed, so I checked on changes to existing code (`OrderFeesHandler` and `_enterprise_fee_summary.html.haml`). They seem like they would be covered well by automated tests already, so I don't think any additional testing needed.

Ready to test :D

drummer83 commented:
Hi @abdellani,

Thanks for this PR! I had a look at it and here are my results.

# Important results
1. At some point during testing I started to see timeouts (error 504) and I couldn't test anymore. It worked in the beginning though. Maybe I loaded the server more and more over time? I have now decided to look at #10413 first, because this might be related? The previous 'enterprise fee summary' report started timing out as well...
2. I didn't see the 'produce' line in the report. It was specified to see it in the results though. Unless there was a decision somewhere to leave it out, we should include that.
Specification:
![image](https://user-images.githubusercontent.com/1790176/218812354-35f825c6-782b-4d61-a2ef-329d14aa4483.png)
Implementation:
![image](https://user-images.githubusercontent.com/1790176/218812441-ef62e6b4-c82d-40ca-b818-5eb2b51913aa.png)

Moving back to In Dev for now. Feel free to move back to Test Ready if you think that everything is ok.

# Less important results (my unfinished notes basically)

## Test cases
- Available for admin and super admin users :heavy_check_mark: 
- Check for regressions on 'Enterprise Fees Summary' report :heavy_check_mark: (see section below)
- All fields described above are present :heavy_check_mark: 
  Note: The field Tax rate name has been added to the report. :exclamation: (probably ok)
- Data matches the description
- Report can be downloaded as a CSV
- Report can be accessed via the UI :heavy_check_mark: 
- Report can be filtered on all fields mentioned above
- Totals row show correctly summed total amounts
- In the UI figures should be rounded to 2 decimal places. In the CSV figures should be rounded to 6 decimal places.
- All fields are translatable

## Check for regressions on 'Enterprise Fees Summary' report
Since parts of this report were moved to new files I checked this report as well.

### Before the PR
Reports list:
![image](https://user-images.githubusercontent.com/1790176/218767434-18cb1d43-3e92-4683-bb7a-a37ec069059e.png)

Reports filters:
![image](https://user-images.githubusercontent.com/1790176/218767505-a69a1c50-a41f-4b9f-9def-b3928af87037.png)

Reports results ([csv BEFORE here](https://github.com/openfoodfoundation/openfoodnetwork/files/10733363/enterprise_fee_summary_20230215_BEFORE.csv)):
![image](https://user-images.githubusercontent.com/1790176/218767648-ce50d02c-458b-4f40-bfcb-8bff1ec71f17.png)

### After the PR
Reports list:
![image](https://user-images.githubusercontent.com/1790176/218769828-edca60f3-ef87-4d34-a32a-820b930db69f.png)

Reports filters:
![image](https://user-images.githubusercontent.com/1790176/218770195-f119d98b-d3b3-4c0c-be56-ae5e7b7a1785.png)

Reports results ([csv AFTER here](https://github.com/openfoodfoundation/openfoodnetwork/files/10733506/enterprise_fee_summary_20230215_AFTER.csv)):
![image](https://user-images.githubusercontent.com/1790176/218770756-16f7514d-c7d4-4973-8e99-fa0c9ccdaadd.png)

### Results
Everything is identical before and after this PR. :+1:

abdellani commented:
Hi @drummer83 

Thank you :pray: for the feedback
> I didn't see the 'produce' line in the report. It was specified to see it in the results though. Unless there was a decision somewhere to leave it out, we should include that.

Sorry, I missed that.
> Note: The field Tax rate name has been added to the report. exclamation (probably ok)

I think this is Ok. https://github.com/openfoodfoundation/openfoodnetwork/issues/6848#issuecomment-1367668919

filipefurtad0 commented:
> I didn't see the 'produce' line in the report. It was specified to see it in the results though. Unless there was a decision somewhere to leave it out, we should include that.

I've missed that as well, when reviewing the spec... I usually pull the code and run the spec to see it running in the browser, which I did not do in this case. Thanks for catching that one Konrad.

abdellani commented:
@lin-d-hop 
At this level, I need help. I'm not sure about how to introduce the `cost of produce` without making the report more complex (and slower?) than what we have now.

mkllnk commented on lib/reporting/reports/enterprise_fee_summary/enterprise_fees_with_tax_report_by_producer.rb:
```diff
@@ -0,0 +1,352 @@
+        end
+
+        def order_permissions
+          return @order_permissions unless @order_permissions.nil?
+
+          @order_permissions = ::Permissions::Order.new(user, ransack_params)
```
```suggestion
          @order_permissions ||= ::Permissions::Order.new(user, ransack_params)
```

mkllnk commented on lib/reporting/reports/enterprise_fee_summary/enterprise_fees_with_tax_report_by_producer.rb:
```diff
@@ -0,0 +1,352 @@
+          @report_line_items ||= Reporting::LineItems.new(order_permissions, params)
+        end
+
+        def supplier_ids_filter(supplier_id)
+          return true if params[:supplier_id_in].blank?
+
+          params[:supplier_id_in].include?(supplier_id)
+        end
```
In Rails, methods returning a boolean value usually end with `?` to mark that. What do you think about this version? (just my opinion)

```suggestion
        def filter_supplier_ids?(supplier_id)
          params[:supplier_id_in].blank? ||
            params[:supplier_id_in].include?(supplier_id)
        end
```

abdellani commented:
Hi @dacook

It looks good to me :)

Thank you very much :pray:

dacook commented:
Great, I've rebased the earlier tweaks, and now ready for review. @mkllnk can you please review my latest addition?

mkllnk reviewed (approved):
:+1:

filipefurtad0 commented:
Apologies - I've posted my **testing notes** on the issue, instead of doing it in this PR: you can [find them here](https://github.com/openfoodfoundation/openfoodnetwork/issues/6844#issuecomment-1505316754) - hence the `feedback needed` label.

dacook commented:
@filipefurtad0 , I've rebased and this is ready to test again.

filipefurtad0 commented:
Hey @dacook @abdellani ,

To test this, it's useful to consider the case in which we have a hub, with three different suppliers (1, 2 and 3).

There are two distinct scenarios tested, corresponding to two order cycles:

- Fees are set up in the **Incoming section**/exchanges of the Order Cycle (left side of the pics) - `order cycle 1`
- Fees are set up in the **Outgoing section**/exchanges of the Order Cycle (left side of the pics) - `order cycle 2`

![image](https://user-images.githubusercontent.com/49817236/235377687-7d4c7e99-8c1e-418f-884f-e27048caa41b.png)

For each of these order cycles, 4 orders were placed:
- order A -> with one item from supplier 1
- order B -> with one item from supplier 2
- order C -> with one item from supplier 3
- order D -> with one item from each supplier (three line items)

As the order cycle coordinator (and hub, in this case), we can render the report, for each corresponding order cycle:

`order cycle 1` **- left**; `order cycle 2` **- right**

![image](https://user-images.githubusercontent.com/49817236/235375478-3766db7e-4f30-4c67-b836-99ab3659ed5d.png)

**Comments**

1) Fee grouping per supplier

It seems that regardless of how the fees are set (incoming vs outgoing sections), the fee grouping per producer is the same. I believe this is incorrect: setting-up fees in the incoming section, allows to selectively apply them to the produce of a given supplier only. This should be correctly be reflected in the totals.

For example, for the report for the order cycle with the fees set as **incoming** (left):
The fees from Red Berries Hub and Wasabi Hub should not appear on the section 1 of the report (corresponding to Kimchi Hub) :x: 
They should (and do appear) on the report for the order cycle with the fees set as **outgoing* (right) :heavy_check_mark: 

2) Cost of produce
The cost of produce Total does not reflect the expected total for a given producer. For supplier 1 (Kimchi Hub), there are items and collected fees on orders A and D, below:

![image](https://user-images.githubusercontent.com/49817236/235377949-2ae35054-3ee1-4afe-acf5-373fe3b6112d.png)

We can see that:
Products total - 2 x (5.00) = 10
Fees total - 2 x (0.5) = 1
Taxes were not applied.

These values are expected in the totals section of the report, for supplier 1 (Kimchi Hub). However, the totals seem to sum up the produce incorrectly. This occurs regardless of how fees are set.

3) Access-rights to this report

Logging in as one of the suppliers, the generated report contains the same information as the version generated by the OC-coordinator/distributor. I'm not sure this is correct, as there may be privacy: suppliers should not be able to access the fees charged by other suppliers. Right?

I'd propose to move this to in dev, to address issues 1) and 2).

Additional note: The set up here is a bit complex and it took me a bit of time. If you feel it is helpful to schedule a chat, let's do so then. In the meantime, I'll try to improve the specs to reflect these scenarios.

dacook commented:
Hi @filipefurtad0 , thanks for your detailed notes, it's really helpful to have your knowledge around this area. In hindsight, I think that we should have developed some test cases like this _before_ starting development, so I might suggest we do that for the [next report](https://github.com/openfoodfoundation/openfoodnetwork/issues/6845).

But as for what next, yes that would be fantastic if you could develop some test cases. I'd like to also add some unit tests to help manage this large report.
But first I think it would be helpful to have a chat about this, so I've sent an invite.

abdellani commented:
@filipefurtad0 
The PR is passing your new tests now
They were really useful, they helped me detect another issue.
Thank you :+1: 

@dacook
I think the only remaining part is the WIP tests. Can you please complete them?

Thank you for adding the cost of produce :pray: 

I think this is a good time to have another review.

dacook commented on spec/system/admin/reports/enterprise_summary_fees/enterprise_summary_fee_with_tax_report_by_producer_spec.rb:
```diff
@@ -77,149 +120,232 @@
     #   1 order cycle has:
     #     - coordinator fees  (20) 1.5% = 0.30, 2.5% = 0.50
-    #     - incoming exchange (15) 1.5% = 0.23, 2.5% = 0.38
+    #     - incoming exchange (15) 1.5% = 0.38, 2.5% = 0.63
```
For "an order with line items from a single supplier (1)" on [line 187](https://github.com/openfoodfoundation/openfoodnetwork/pull/10304/commits/3cf03441ef5bc1ce4b0376933087539c47f7b1b3#diff-9b73e96b812e7861f7876d3d5cc0ccaa4dba276d3645ef334ea2d29e34f1fa8aR187) below, we see the old numbers:

```ruby
          "5.81",  # Tax     : (0.30 + 0.50) + (0.23 + 0.38) + (0.15 + 0.25) + (1.50 + 2.50)
```

But for "an order with line items from a single supplier (2)" on line [326](https://github.com/openfoodfoundation/openfoodnetwork/pull/10304/commits/3cf03441ef5bc1ce4b0376933087539c47f7b1b3#diff-9b73e96b812e7861f7876d3d5cc0ccaa4dba276d3645ef334ea2d29e34f1fa8aR326) we see the updated numbers:
```ruby
          "4.21",  # Tax     : (0.30 + 0.50) + (0.38 + 0.63) + (0.15 + 0.25) + 2
```

I can't see why. Could the comment be updated to make it clearer perhaps?

filipefurtad0 replied:
Hum, that does not seem right - thanks for spotting that @dacook . I'll proceed as before: pull this branch and add the necessary corrections to the comments.

dacook reviewed (commented):
Thanks Filipe and Mohamed, epic! 💪 
The system spec really is quite a beast now, so I appreciate your work on this. 

For one commit, I have some suggestions on how this could be arranged simpler and more performant, @abdellani can you please take a look and try it out if it looks right?

@filipefurtad0 for the spec, I think it would be helpful to have some more introductory comments for each `context`. But I realise this was written a while ago now, so don't expect you to do that now. 
However I do have one query about something I wasn't able to explain myself.

I will see if I have time to try adding some unit tests this week. I think this could be a really helpful framework to have in place for next time we need to work on the report (once there's specs, we have to keep them up to date ;)). 
But if we have it working already with system specs, I think it's optional and shouldn't hold back this PR!

abdellani commented on lib/reporting/reports/enterprise_fee_summary/enterprise_fees_with_tax_report_by_producer.rb:
```diff
@@ -119,22 +119,47 @@ def join_tax_rate
                   order: order
                 }
               end
+              .filter do |hash|
+                hash[:enterprise_fee_id].present?
+              end
```
done

abdellani commented on lib/reporting/reports/enterprise_fee_summary/enterprise_fees_with_tax_report_by_producer.rb:
```diff
@@ -119,22 +119,47 @@ def join_tax_rate
+                                     end,
+                  supplier_id: line_item.supplier.id,
                   order: order
                 }
```
instead, I suggest filtering and then mapping each item.

dacook replied:
Ok, I think my suggestion of merging is cleaner (no need to re-declare the `tax_rate_id` and `enterprise_fee_id` that are copied across). Actually, it might have been even cleaner and more efficient by editing the line_item hash directly (eg `line_item[:supplier_id] = line_item.supplier.id`)

But this works. Hopefully it's not too hard to optimise in the future if we need to.

abdellani replied:
Sorry, I misunderstood your earlier comment. 

I was mapping before filtering, but I fixed that now.
 
Also, you're right, it'll be better to avoid using mapping and to modify the line_item objects directly.
I tried to do that, but it seems not possible. Please let me know what you think. 
```ruby
        def join_supplier
          proc do |item|
            order = item[:order]
            enterprise_fees_per_variant = enterprise_fees_per_variant(order)
            filtered_line_items(order)
              .filter do |line_item|
                item[:enterprise_fee_id].in?(
                  enterprise_fees_per_variant[line_item.variant]
                )
              end
              .each do |line_item|
                line_item[:tax_rate_id]= item[:tax_rate_id]
                line_item[:enterprise_fee_id]= item[:enterprise_fee_id]
                line_item[:supplier_id]= line_item.supplier.id
                line_item[:order]= order
              end
          end
        end
``` 
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/16113997/df166c56-eaff-4066-81fd-2675e5c8f4e3)

dacook replied:
Ah, thanks for trying. And sorry, I forgot that in this case, we're iterating over LineItems, not hashes. 
And other methods are relying on this one to aggregate some data into a hash, so we can't easily change this.

So lets move one :)

filipefurtad0 commented:
Thanks for pinging me @abdellani @dacook (and for generating those [files here](https://github.com/openfoodfoundation/openfoodnetwork/pull/10304#pullrequestreview-1478408384) :muscle: )

I had done something local, but still - I've missed a detail... My approach for this spec was:

- to have order 1, oc1 with one line item from supplier 1 -> check the fees
- to have order 2, oc2 with one line item from supplier 2 -> check the fees
- to have order 3, oc3 with one line item for from supplier 1 + another line item from supplier 2 -> the fees should be the same for the respective suppliers; i.e., having more suppliers on the same OC/order should not interfere with the respective fees (for each producer).

However, the spec I've written was not selective enough and I believe I've missed some incorrect values on the order total (sorry for that). As far as I understand, the values marked in red :red_circle: should be the same as those marked in :green_circle: (we can easily verify that the values above these markings are identical, so I see no reason for the totals not to match):

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/9cb8f617-ee5f-41fb-a496-066052aef73f)

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/320f3586-6c04-4161-8c82-1a85632e5931)

I believe I can make the spec work as intended if I just make better use of the filtering function (now it's working - awesome! :muscle:). I'll edit this message with the correct commit.

dacook commented:
Thanks Filipe, just confirming that the next step is to fix the pending spec you've created.

dacook commented:
Rebasing..

rioug commented on spec/system/admin/reports/enterprise_summary_fees/enterprise_summary_fee_with_tax_report_by_producer_spec.rb:
```diff
@@ -0,0 +1,216 @@
+    it 'generates the report' do
+      login_as admin
+      visit admin_reports_path
+      click_on I18n.t("admin.reports.enterprise_fees_with_tax_report_by_producer")
```
The preference is to use the translated string for system specs

rioug commented on spec/system/admin/reports/enterprise_summary_fees/enterprise_summary_fee_with_tax_report_by_producer_spec.rb:
```diff
@@ -0,0 +1,216 @@
+    it 'generates the report' do
+      login_as admin
+      visit admin_reports_path
+      click_on I18n.t("admin.reports.enterprise_fees_with_tax_report_by_producer")
```
Same as before, the preference is to use the translated string for system specs

rioug commented on spec/system/admin/reports/enterprise_summary_fees/enterprise_summary_fee_with_tax_report_by_producer_spec.rb:
```diff
@@ -189,163 +235,363 @@
+
+          login_as distributor_owner
+          visit admin_reports_path
+          click_on I18n.t("admin.reports.enterprise_fees_with_tax_report_by_producer")
```
The preference is to use the translated string in system specs

rioug commented on spec/system/admin/reports/enterprise_summary_fees/enterprise_summary_fee_with_tax_report_by_producer_spec.rb:
```diff
@@ -189,163 +235,363 @@
+          before do
+            login_as distributor_owner
+            visit admin_reports_path
+            click_on I18n.t("admin.reports.enterprise_fees_with_tax_report_by_producer")
```
The preference is to use the translated string in system specs

rioug commented on spec/system/admin/reports/enterprise_summary_fees/enterprise_summary_fee_with_tax_report_by_producer_spec.rb:
```diff
@@ -77,149 +120,232 @@
+        it "should filter by fee name" do
+          login_as distributor.owner
+          visit admin_reports_path
+          click_on I18n.t("admin.reports.enterprise_fees_with_tax_report_by_producer")
```
The preference is to use the translated string in system specs

rioug commented on spec/system/admin/reports/enterprise_summary_fees/enterprise_summary_fee_with_tax_report_by_producer_spec.rb:
```diff
@@ -602,76 +591,153 @@
+        pending("test case (3), see #10797")
+        login_as distributor_owner
+        visit admin_reports_path
+        click_on I18n.t("admin.reports.enterprise_fees_with_tax_report_by_producer")
```
The preference is to use the translated string in system specs

rioug commented on spec/system/admin/reports/enterprise_summary_fees/enterprise_summary_fee_with_tax_report_by_producer_spec.rb:
```diff
@@ -309,6 +309,37 @@
+          before do
+            login_as distributor_owner
+            visit admin_reports_path
+            click_on I18n.t("admin.reports.enterprise_fees_with_tax_report_by_producer")
```
The preference is to use the translated string in system specs

rioug reviewed (approved):
This still goes over my head, but as far as I can tell the code looks fine. 
I just left a few comments about using the translated string in system specs, I would be good to clean that up, but nothing blocking.

dacook reviewed (approved):
Great, this looks way simpler, well done!

drummer83 commented:
Hi team,

I had a look at this report and I would like to share my first thoughts.

### Test setup
- I used an enterprise which charges VAT.
  - VAT is 20 % in AB (Alberta), 10 % in BC (British Columbia), 0 % outside.
  - VAT is included in price.
- Product price is 100 $.
- A packing fee of 10 $ applies per order (incl. VAT tax).
- A transaction fee of 2 $ applies. 

### Test carts
1. 1 item of product, delivery outside AB and BC -> no taxes apply
2. 1 item of product, delivery to BC -> 10 % tax applies
3. 1 item of product, delivery to AB -> 20 % tax applies

### Test results
Here is the report:
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/4d1390ba-7da8-4b2d-91da-5829874e756d)

- The rows of the report all seem correct to me: `Total excl. tax + tax = Total incl. tax` :heavy_check_mark: 
- The row TOTALS seem incorrect to me (see image above) :x: 
- The row 'Cost of produce' is only displayed when 'Summary Row' is active in the report rendering options. This seems incorrect to me :x: 

## Conclusion
Before testing more complex setups I think we should get this example right. So I will move to In Dev.
I have read all of threads above, but in case I am missing something in my findings, just let me know, please. Thanks!

Pinging @abdellani, @filipefurtad0, @lin-d-hop.

lin-d-hop commented:
Hi @drummer83 
If you are looking at this through the lens of "how much do I have to pay my suppliers" is it still wrong?

drummer83 commented:
Yes, I would think so? 🤔

lin-d-hop commented:
Ok, I'll have a proper look tomorrow. Something's not right. Just to notes enterprise fees and transaction fees should be ignored on this report.

abdellani commented:
@drummer83 

I don't know about the configuration, but from the screenshot, it looks like the three rows refer to the same Enterprise free.
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/16113997/2827370f-0280-4cc9-8d5a-ccf5445afff0)

Can you please confirm by checking the order cycle?

drummer83 commented:
> Just to notes enterprise fees and transaction fees should be ignored on this report.

Well, that's surprising! I thought enterprise fees should be considered in an enterprise fees report? :wink: 
Transaction fees are ignored and I think that's ok. :+1:

drummer83 commented:
> it looks like the three rows refer to the same Enterprise free

Yes, @abdellani, it's the same enterprise fee, but for three different orders with three different tax rates (due to delivery addresses in three different tax zones). So I think it's ok to list it three times in the report. :+1:

abdellani commented:
> Yes, @abdellani, it's the same enterprise fee, but for three different orders with three different tax rates (due to delivery addresses in three different tax zones). So I think it's ok to list it three times in the report. :+1: 


I don't see a problem with repetition, since each row is connected to a different tax rate.

I wanted to confirm because if they were related to the same order, the total would be correct (284.24).

I'll inspect the issue tomorrow.

abdellani commented:
I fixed the issues and implemented the scenario illustrated by @drummer83 ( added and included taxes).

The `added taxes` test is passing without any issues.
I'm stuck with `included taxes`. For some reason, it's showing taxes with negative values (refund).
The problem is not on the report level, but maybe in the configuration of the test.
@filipefurtad0 can you please double-check the configuration?

abdellani commented:
The `included taxes` test was failing because I was trying to apply included taxes on an order where the tax zone and shipping address belong to two different countries. I was getting refunds on that order (negative taxes).

I fixed the issue by using only one country on the new tests.

In the new scenario, we have 2 orders that belong to the same country but belong to two different states. 

# Added taxes

| Distributor | Producer | Producer Tax Status | Order Cycle | Name | Type | Owner | Tax Category | Tax Rate Name | Tax Rate | Total excl. tax | Tax | Total incl. tax |
|---|---|---|---|---|---|---|---|---|---|---|---|---|
| Distributor | Supplier1 | Yes | oc1 | Adminstration | admin | Distributor | tax_category | State | 0.015| 20.0 | 0.3 | 20.3 |
|Distributor| Supplier1| Yes| oc1| Adminstration| admin| Distributor| tax_category| Country| 0.025| 40.0| 1.0| 41.0| 
|Distributor| Supplier1| Yes| oc1| Transport| transport| Supplier1| tax_category| State| 0.015| 15.0| 0.23| 15.23| 
|Distributor| Supplier1| Yes| oc1| Transport| transport| Supplier1| tax_category| Country| 0.025| 30.0| 0.76| 30.76| 
|Distributor| Supplier1| Yes| oc1| Packing| packing| Distributor| tax_category| State| 0.015| 10.0| 0.15| 10.15| 
|Distributor| Supplier1| Yes| oc1| Packing| packing| Distributor| tax_category| Country| 0.025| 20.0| 0.5| 20.5|
|Distributor| Supplier1| Yes| oc1| Adminstration| admin| Distributor| tax_category| Another State Tax| 0.02| 20.0| 0.4| 20.4| 
|Distributor| Supplier1| Yes| oc1| Transport| transport| Supplier1| tax_category| Another State Tax| 0.02| 15.0| 0.3| 15.3| 
|Distributor| Supplier1| Yes| oc1| Packing| packing| Distributor| tax_category| Another State Tax| 0.02| 10.0| 0.2| 10.2| 
|Distributor| Supplier1| Yes| oc1| Cost of produce| line items| Supplier1|||| 150.0| 6.25| 156.25| 
||||||||||TOTAL| 240.0| 10.09| 250.09|

# included taxes
| Distributor | Producer | Producer Tax Status | Order Cycle | Name | Type | Owner | Tax Category | Tax Rate Name | Tax Rate | Total excl. tax | Tax | Total incl. tax |
|---|---|---|---|---|---|---|---|---|---|---|---|---|
|Distributor| Supplier1| Yes| oc1| Adminstration| admin| Distributor| tax_category| State| 0.015| 19.21| 0.3| 19.51| 
|Distributor| Supplier1| Yes| oc1| Adminstration| admin| Distributor| tax_category| Country| 0.025| 38.33| 0.98| 39.31| 
|Distributor| Supplier1| Yes| oc1| Transport| transport| Supplier1| tax_category| State| 0.015| 14.41| 0.22| 14.63| 
|Distributor| Supplier1| Yes| oc1| Transport| transport| Supplier1| tax_category| Country| 0.025| 28.75| 0.74| 29.49| 
|Distributor| Supplier1| Yes| oc1| Packing| packing| Distributor| tax_category| State| 0.015| 9.61| 0.15| 9.76| 
|Distributor| Supplier1| Yes| oc1| Packing| packing| Distributor| tax_category| Country| 0.025| 19.17| 0.48| 19.65|
|Distributor| Supplier1| Yes| oc1| Adminstration| admin| Distributor| tax_category| Another State Tax| 0.02| 19.12| 0.39| 19.51| 
|Distributor| Supplier1| Yes| oc1| Transport| transport| Supplier1| tax_category| Another State Tax| 0.02| 14.34| 0.29| 14.63| 
|Distributor| Supplier1| Yes| oc1| Packing| packing| Distributor| tax_category| Another State Tax| 0.02| 9.56| 0.2| 9.76|
|Distributor| Supplier1| Yes| oc1| Cost of produce| line items| Supplier1|||| 143.88| 6.12| 150.0| 
||||||||||TOTAL| 230.13| 9.87| 240.0|

dacook reviewed (commented):
Looks good 👍 
The state setup doesn't look very fun :(

I have an idea to reduce duplication in the spec, and [starting doing it](https://github.com/dacook/openfoodnetwork/commits/add_enterprise_fees_w_tax_report_by_producer), but couldn't quite finish off yet. 

So I'd like to try and finish this tomorrow morning if that's ok with you?

lin-d-hop commented:
> > Just to notes enterprise fees and transaction fees should be ignored on this report.
> 
> Well, that's surprising, @lin-d-hop! I thought enterprise fees should be considered in an enterprise fees report? wink Transaction fees are ignored and I think that's ok. +1

They will be included by order, but for the producer report we only include fees relevant to the producer. If the user wants see all order level fees then it is necessary to consider the report per order. Transaction fees can't be sorted by producer :)

lin-d-hop commented:
> The row TOTALS seem incorrect to me (see image above) :x:
> The row 'Cost of produce' is only displayed when 'Summary Row' is active in the report rendering options. This seems incorrect to me :x:

Hi @drummer83, I agree with both of these comments.

Certainly the Cost of Produce should display even if the Summary Row is not shown.

I agree the Totals look odd. If these are the packing fees for 3 orders then you'd expect all three idn the totals. But if they are order level Enterprise Fees, as a flat fee per order, then they shouldn't appear in this report at all.

The intended scenario is that Coordinator Fees are left off this report, but fees added at OC level per producer should be added. However it seems this requirement probably has been lost by now if I am saying it here :( 
If this was the case then it would be correct to add for different tax zone as separate line items as in the case above, but you'd expect the total to reflect all three tax zones which it doesn't in the example you shared.

Would you test around this case (with coordinator fees and producer level fees) just to get an idea of how far from this requirement we actually now? Thanks!

filipefurtad0 commented:
> But if they are order level Enterprise Fees, as a flat fee per order, then they shouldn't appear in this report at all

I've understood this as we've agreed to ship as is, on this regard (last week, delivery circle). Should this behavior be changed, within the scope of this PR? (specs don't reflect this, currently)

A more general question: does it make sense to be able to add order-based calculators fees, as a supplier (incoming exchange of the OC)?  

> The intended scenario is that Coordinator Fees are left off this report

Did not understand this in this way so far. In that case, should Distributor fees (added in the Outgoing exchange) appear on this report?

lin-d-hop commented:
> A more general question: does it make sense to be able to add order-based calculators fees, as a supplier (incoming exchange of the OC)?

These are a tricky case. For a hub with multiple producers there is no way to add flat rate per order fees accurately when sorting by producer, so.... no. I'm not sure these fees are ever used in this way so I'd say leave off and wait for complaints.

> Did not understand this in this way so far. In that case, should Distributor fees (added in the Outgoing exchange) appear on this report?

Nope. These are post-packing fees so should be left off this report (another way to look at it is a pre-packing report)

> I've understood this as we've agreed to ship as is, on this regard (last week, delivery circle). Should this behavior be changed, within the scope of this PR? (specs don't reflect this, currently)

ok

filipefurtad0 commented:
Ok, so I think a good summary would be:
- Coordinator fees -> do not display
- Distributor fees -> do not display
- Supplier fees, with per-order calculator -> sort of an edge case, ideally do not display
- Supplier fees, with all line-item-based calculators -> display

jibees commented:
Shouldn't this one be `In Dev`?

abdellani commented:
the pending work was done on this [PR](https://github.com/openfoodfoundation/openfoodnetwork/pull/11013)
